File: hw/vargas/u1/hw16IncomeTax/IncomeTaxPY/incomeTaxCalculation.py
def CalculateTaxableBase(MonthlySalary: float) -> float:
    AnnualIncome = MonthlySalary * 12
    AnnualIessContribution = AnnualIncome * 0.0945
    return AnnualIncome - AnnualIessContribution

def GetMaximumDiscount(NumberOfFamilyDeficits: int) -> float:
    if NumberOfFamilyDeficits == 0:
        return 1005.87
    elif NumberOfFamilyDeficits == 1:
        return 1293.26
    elif NumberOfFamilyDeficits == 2:
        return 1580.65
    elif NumberOfFamilyDeficits == 3:
        return 2011.74
    elif NumberOfFamilyDeficits == 4:
        return 2442.83
    else:
        return 2873.92

def CalculateFinalDiscount(AnnualPersonalExpenses: float, NumberOfFamilyDeficits: int) -> float:
    CalculatedDiscount = AnnualPersonalExpenses * 0.18
    MaximumDiscount = GetMaximumDiscount(NumberOfFamilyDeficits)
    return min(CalculatedDiscount, MaximumDiscount)

def CalculateTaxDue(TaxBase: float) -> float:
    BasicFraction = 0.0
    BasicTax = 0.0
    PercentageSurplus = 0.0

    if TaxBase <= 12081:
        BasicFraction = 0
        BasicTax = 0
        PercentageSurplus = 0.0
    elif TaxBase <= 15387:
        BasicFraction = 12081
        BasicTax = 0
        PercentageSurplus = 0.05
    elif TaxBase <= 19978:
        BasicFraction = 15387
        BasicTax = 165
        PercentageSurplus = 0.10
    elif TaxBase <= 26422:
        BasicFraction = 19978
        BasicTax = 624
        PercentageSurplus = 0.12
    elif TaxBase <= 34770:
        BasicFraction = 26422
        BasicTax = 1398
        PercentageSurplus = 0.15
    elif TaxBase <= 46089:
        BasicFraction = 34770
        BasicTax = 2650
        PercentageSurplus = 0.20
    elif TaxBase <= 61359:
        BasicFraction = 46089
        BasicTax = 4914
        PercentageSurplus = 0.25
    elif TaxBase <= 81817:
        BasicFraction = 61359
        BasicTax = 8731
        PercentageSurplus = 0.30
    elif TaxBase <= 108810:
        BasicFraction = 81817
        BasicTax = 14869
        PercentageSurplus = 0.35
    else:
        BasicFraction = 108810
        BasicTax = 24316
        PercentageSurplus = 0.37

    Surplus = TaxBase - BasicFraction
    TaxDue = BasicTax + (Surplus * PercentageSurplus)
    return TaxDue


def IncomeTaxCalculation(MonthlySalary: float, AnnualPersonalExpenses: float, NumberOfFamilyDeficits: int) -> float:
    
    TaxBase = CalculateTaxableBase(MonthlySalary)
    TaxDue = CalculateTaxDue(TaxBase)

    if TaxDue == 0.0:
        return 0.0

    FinalDiscount = CalculateFinalDiscount(AnnualPersonalExpenses, NumberOfFamilyDeficits)
    TaxToPay = TaxDue - FinalDiscount

    return max(0.0, TaxToPay)

File: hw/vargas/u1/hw16IncomeTax/IncomeTaxJS/incomeTaxCalculation.js
function _calculateTaxableBase(monthlySalary) {
    const annualIncome = monthlySalary * 12;
    const annualIESSContribution = annualIncome * 0.0945;
    return annualIncome - annualIESSContribution;
}

function _getMaximumDiscount(numberOfFamilyDeficits) {
    switch (numberOfFamilyDeficits) {
        case 0: return 1005.87;
        case 1: return 1293.26;
        case 2: return 1580.65;
        case 3: return 2011.74;
        case 4: return 2442.83;
        case 5:
        default:
            return 2873.92;
    }
}

function _calculateFinalDiscount(annualPersonalExpenses, numberOfFamilyDeficits) {
    const calculatedDiscount = annualPersonalExpenses * 0.18;
    const maximumDiscount = _getMaximumDiscount(numberOfFamilyDeficits);
    return Math.min(calculatedDiscount, maximumDiscount);
}

function _calculateTaxDue(taxBase) {
    let basicFraction = 0;
    let basicTax = 0;
    let percentageSurplus = 0;

    if (taxBase <= 12081) {
        basicFraction = 0;
        basicTax = 0;
        percentageSurplus = 0;
    } else if (taxBase <= 15387) {
        basicFraction = 12081;
        basicTax = 0;
        percentageSurplus = 0.05;
    } else if (taxBase <= 19978) {
        basicFraction = 15387;
        basicTax = 165;
        percentageSurplus = 0.10;
    } else if (taxBase <= 26422) {
        basicFraction = 19978;
        basicTax = 624;
        percentageSurplus = 0.12;
    } else if (taxBase <= 34770) {
        basicFraction = 26422;
        basicTax = 1398;
        percentageSurplus = 0.15;
    } else if (taxBase <= 46089) {
        basicFraction = 34770;
        basicTax = 2650;
        percentageSurplus = 0.20;
    } else if (taxBase <= 61359) {
        basicFraction = 46089;
        basicTax = 4914;
        percentageSurplus = 0.25;
    } else if (taxBase <= 81817) {
        basicFraction = 61359;
        basicTax = 8731;
        percentageSurplus = 0.30;
    } else if (taxBase <= 108810) {
        basicFraction = 81817;
        basicTax = 14869;
        percentageSurplus = 0.35;
    } else {
        basicFraction = 108810;
        basicTax = 24316;
        percentageSurplus = 0.37;
    }

    const surplus = taxBase - basicFraction;
    const taxDue = basicTax + (surplus * percentageSurplus);
    return taxDue;
}

export function incomeTaxCalculation(monthlySalary, annualPersonalExpenses, numberOfFamilyDeficits) {
    
    const taxBase = _calculateTaxableBase(monthlySalary);
    const taxDue = _calculateTaxDue(taxBase);

    if (taxDue === 0.0) {
        return 0.0;
    }

    const finalDiscount = _calculateFinalDiscount(annualPersonalExpenses, numberOfFamilyDeficits);
    const taxToPay = taxDue - finalDiscount;

    return Math.max(0.0, taxToPay);
}